File: Ex1_LoopScheduling/saxpy/cpp/yourwork/saxpy_yourwork.cpp
#include <iostream>
#include <vector>

#define ORD 1 << 27 // size of array

using namespace std;

int main()
{
    // instantiate vectors and attribute values
    vector<float> x, y;
    float r, elmnt;
    srand(time(NULL)); // random number generator seed
    for (int i = 0; i < ORD; i++)
    {
        r = rand() % RAND_MAX;         // get a random integer
        elmnt = float(r) / (RAND_MAX); // normalize it
        x.push_back(elmnt);            // x is noise
        y.push_back(1.0);              // y is just ones
    }
    float a = 2.0; // scalar multiplier (y -> a*x + y)

    // perform SAXPY
    for (int i = 0; i < x.size(); i++)
    {
        y[i] = a * x[i] + y[i];
    }
    
    // goodbye
    return 0;
}

File: Ex1_LoopScheduling/saxpy/fortran/yourwork/saxpy_yourwork.f90
program saxpy_yourwork
    use, intrinsic :: iso_fortran_env
    implicit none
    integer, parameter :: sp = REAL32   ! single precision
    integer, parameter :: dp = REAL64   ! double precision
    integer, parameter :: ORD = 2**27   ! arrays size
    real(sp), dimension(:), allocatable :: x, y     ! arrays
    real(sp) :: a   ! scalar constant    
    integer :: i    ! integer helper

    ! allocate memory space
    allocate(x(ORD))
    allocate(y(ORD))
    
    ! initialize random number generator, set values of x, y, a
    call random_seed()
    call random_number(x)   ! x is noise
    y = 1.0_sp              ! y is just ones
    a = 2.0_sp              ! a is the scalar y -> a*x + y

    ! perform SAXPY
    do i = 1, ORD
        y(i) = a*x(i) + y(i)
    enddo

    ! clean up
    deallocate(x,y)

end program saxpy_yourwork
